data_processors/lims/services.py

```diff
@@ -130,3 +137,142 @@ def __csv_column_to_field_name(column_name: str) -> str:
     return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
+
+@transaction.atomic
+def persist_labmetadata(df: pd.DataFrame, rewrite: bool = False) -> Dict[str, int]:
+    """
+    Persist labmetadata from a pandas dataframe into the db
+    :param df: dataframe to persist
+    :param rewrite: whether we are rewriting the data
+    :return: result statistics - count of updated, new and invalid LabMetadata rows
+    """
+    df_starting_len = len(df.index)
+    logger.info(f"Start processing LabMetadata data")
+    if rewrite:
+        # Delete all rows first
+        logger.info("REWRITE MODE: Deleting all existing records")
+        LabMetadata.objects.all().delete()
+
+    # clean the DF and reset the index
+    df = df.applymap(_clean_datacell)
+    df = df.drop_duplicates()
+    df = _remove_rows_with_empty_required_cols(df)
+    df = df.reset_index(drop=True)
+
+    # Below code assumes a labmetadata's unique identifier is a tuple of library_id and sample_name
+
+    # Assemble a df of entries to update by first querying for existing labmetadatas 
+    libids_samplenames = list(zip( df['library_id'].tolist(),  df['sample_name'].tolist() ))
+    query = reduce(operator.or_, (Q(library_id=l, sample_name=s) for l, s in libids_samplenames))
+    existing_labmetadatas = LabMetadata.objects.filter(query)
+    existing_identifier_tuples = tuple((lm.library_id, lm.sample_name) for lm in existing_labmetadatas)
+    df_to_update =  df[pd.Series(list(zip(df['library_id'], df['sample_name']))).isin(existing_identifier_tuples)]
+    
+    # Assemble df of entries to create.
+    df_to_create = df.append(df_to_update)
+    df_to_create = df_to_create[~df_to_create.index.duplicated(keep=False)]
+
+    # Create labmetadata ojbects
+    try:
+        instances_insert = _make_labmetadata_instances_from_dataframe(df_to_create)
+        rows_created = LabMetadata.objects.bulk_create(instances_insert,batch_size=100)
+    except Exception as e:
+        logger.error("Error bulk creating objects! No new rows created")
+        logger.error(e)   
+        logger.debug(traceback.format_exc())
+        rows_created = []
+        raise 
+
+    # Update labmetadatas
+    rows_updated = 0
+    for row in df_to_update.itertuples():
+        rows_updated += _update_labmetadata_instance(row)
+
+    logger.debug("Updated " + str((rows_updated)))
+    logger.debug("New " + str(len(rows_created)))
+    logger.debug("Invalid " + str(len(df.index) - len(rows_created) - rows_updated ))
+
+    return {
+        'labmetadata_row_update_count':rows_updated,
+        'labmetadata_row_new_count': len(rows_created),
+        'labmetadata_row_invalid_count': df_starting_len - len(rows_created) - rows_updated
+    }
+
+def _clean_datacell(value):
+    #python NaNs are != to themselves
+    if value == '-' or value == np.nan or value != value:
+        value = ''
+    if (isinstance(value,str) and value.strip() == ''):
+        value = ''
+    return value
```

Comment:
Perhaps change the order?
```
if isInstance(value, str):
    value = value.strip()  # I think we could get rid of surrounding whitespace for all values, no?
if value == '-' or value == np.nan or value != value:
    value = ''
return value
```
This way you'd catch cases where a dash is surrounded by whitespace.